File: solutions/typescript/question3.ts
/**
 * Q3 — Fazenda de execução de jobs.
 *
 * Abra questions/q3_build_farm/LEVEL_1.md e implemente somente o nível atual.
 * Preserve todas as assinaturas públicas.
 */
export class BuildFarmService {
  addWorker(workerId: string, capacity: number): boolean {
    return false;
  }

  submitJob(timestamp: number, jobId: string, workerId: string, cost: number): boolean {
    return false;
  }

  completeJob(timestamp: number, jobId: string): boolean {
    return false;
  }

  getWorkerLoad(timestamp: number, workerId: string): number | null {
    return null;
  }

  transferJob(timestamp: number, jobId: string, targetWorkerId: string): boolean {
    return false;
  }

  topWorkers(timestamp: number, n: number): string[] {
    return [];
  }

  addDependency(timestamp: number, jobId: string, prerequisiteJobId: string): boolean {
    return false;
  }

  readyJobs(timestamp: number, workerId: string): string[] {
    return [];
  }

  drainWorker(timestamp: number, sourceWorkerId: string, targetWorkerIds: string[]): boolean {
    return false;
  }

  getJobHistory(timestamp: number, jobId: string): string[] {
    return [];
  }
}

File: solutions/python/question3.py
"""Q3 — Fazenda de execução de jobs.

Abra questions/q3_build_farm/LEVEL_1.md e preserve as assinaturas públicas.
"""


class BuildFarmService:
    def add_worker(self, worker_id: str, capacity: int) -> bool:
        return False

    def submit_job(
        self, timestamp: int, job_id: str, worker_id: str, cost: int
    ) -> bool:
        return False

    def complete_job(self, timestamp: int, job_id: str) -> bool:
        return False

    def get_worker_load(self, timestamp: int, worker_id: str) -> int | None:
        return None

    def transfer_job(
        self, timestamp: int, job_id: str, target_worker_id: str
    ) -> bool:
        return False

    def top_workers(self, timestamp: int, n: int) -> list[str]:
        return []

    def add_dependency(
        self, timestamp: int, job_id: str, prerequisite_job_id: str
    ) -> bool:
        return False

    def ready_jobs(self, timestamp: int, worker_id: str) -> list[str]:
        return []

    def drain_worker(
        self, timestamp: int, source_worker_id: str, target_worker_ids: list[str]
    ) -> bool:
        return False

    def get_job_history(self, timestamp: int, job_id: str) -> list[str]:
        return []
